lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -82,6 +127,169 @@ struct ConcatOpConversion : public OpConversionPattern<ConcatOp> {
+  LogicalResult
+  matchAndRewrite(ReduceOrOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Type resultType = typeConverter->convertType(op.getInput().getType());
+    Value zero = rewriter.create<hw::ConstantOp>(op->getLoc(), resultType, 0);
+
+    rewriter.replaceOpWithNewOp<comb::ICmpOp>(op, comb::ICmpPredicate::ne,
+                                              adaptor.getInput(), zero);
+    return success();
+  }
+};
+
+struct ReduceXorOpConversion : public OpConversionPattern<ReduceXorOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(ReduceXorOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    rewriter.replaceOpWithNewOp<comb::ParityOp>(op, adaptor.getInput());
+    return success();
+  }
+};
+
+struct BoolCastOpConversion : public OpConversionPattern<BoolCastOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(BoolCastOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (cast<UnpackedType>(op.getInput().getType())
+            .castToSimpleBitVectorOrNull()) {
+      Type resultType = typeConverter->convertType(op.getInput().getType());
+      Value zero = rewriter.create<hw::ConstantOp>(op->getLoc(), resultType, 0);
+
+      rewriter.replaceOpWithNewOp<comb::ICmpOp>(op, comb::ICmpPredicate::ne,
+                                                adaptor.getInput(), zero);
+      return success();
+    }
+
+    return success();
+  }
+};
+
+struct NotOpConversion : public OpConversionPattern<NotOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(NotOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Type resultType =
+        ConversionPattern::typeConverter->convertType(op.getResult().getType());
+    Value max = rewriter.create<hw::ConstantOp>(op.getLoc(), resultType, -1);
+
+    rewriter.replaceOpWithNewOp<comb::XorOp>(op, adaptor.getInput(), max);
+    return success();
+  }
+};
+
+template <typename SourceOp, typename TargetOp>
+struct BinaryOpConversion : public OpConversionPattern<SourceOp> {
+  using OpConversionPattern<SourceOp>::OpConversionPattern;
+  using OpAdaptor = typename SourceOp::Adaptor;
+
+  LogicalResult
+  matchAndRewrite(SourceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (isa<DivOp>(op) && isSignedType(op)) {
+      rewriter.replaceOpWithNewOp<comb::DivSOp>(op, adaptor.getLhs(),
+                                                adaptor.getRhs(), false);
+      return success();
+    }
+    if (isa<ModOp>(op) && isSignedType(op)) {
+      rewriter.replaceOpWithNewOp<comb::ModSOp>(op, adaptor.getLhs(),
+                                                adaptor.getRhs(), false);
+      return success();
+    }
```

Comment:
This special case isn't particularly elegant. Maybe we should create another pattern with 3 template arguments so that the signed version of an op can be given as an argument.

Reply:
Get it. I'll modify it.

Reply:
> This special case isn't particularly elegant. Maybe we should create another pattern with 3 template arguments so that the signed version of an op can be given as an argument.

How about this?

Reply:
Very nice!

Reply:
> Very nice!

Thanks so much for pointing out the code inadequacies!!! :heart: